Automattic/newspack-newsletters
# feat: refactor how we store lists configuration

### All Submissions:






### Changes proposed in this Pull Request:



This PR refactors how we store the lists configuration.

Instead of saving the lists configuration as an option in the database, we now create a custom post for each list. We are using the same post type used to store "local" lists, but it has been extended so it now also serves to store "remote" lists.

"local" lists are those lists that are created via wp-admin. When we create a local list, the plugin creates a tag (or group) in the ESP
"remote" lists are the lists that are dynamically fetched from the ESP. Those lists are always managed in the ESP. Users can set an alternate title and description for them, but this information is used only in the site and is not sent to the ESP.

Everytime a user visits the admin page:
* we fetch all the lists from the ESP
* we get a post that represents that list in the local database (the custom post). If the post doesn't exist yet, we create it
* We store the ID the list has in the ESP as a metadata called "remote_id"
* We allow lists to be enabled or disabled, represented by the post status publish or draft

Test migration:
* On master, select a provider and set up a couple of lists
* Make sure the migration flag does not exist `wp option delete _newspack_newsletters_lists_migrated`
* Checkout this branch and confirm the lists configuration is still the same
* Repeat with a couple of providers and with some local lists in the mix

Local lists are still treated the same. The difference is that we now handle the activation/deactivation by changing the post status. Same thing we do with the remote lists.

We don't use the configuration options we used to have any more.

### How to test the changes in this Pull Request:

Test everything related to list management:

1. Set up the provider as Mailchimp
2. Enable a couple of lists and customize their titles and descriptions
3. Set up a subscription block and confirm it works
4. Check that both the Newsletters > Settings UI and the Newspack wizard UI work the same way
5. Create a local list and confirm it works as expected
6. Try activating and deactivating different lists
7. Change to Active Campaign
8. Do all the tests again, inlcuding local lists
9. Switch back to Mailchimp
10. Confirm that the title customizations and the local lists you had created before a back
11. Test Campaign Monitor and confirm it works (provider that don't support local lists)

### Other information:
